pkg/platform/kube/monitoring/test/function_test.go

```diff
@@ -0,0 +1,312 @@
+package test
+
+import (
+	"encoding/base64"
+	"fmt"
+	"testing"
+	"time"
+
+	"github.com/nuclio/nuclio/pkg/common"
+	"github.com/nuclio/nuclio/pkg/functionconfig"
+	"github.com/nuclio/nuclio/pkg/platform"
+	"github.com/nuclio/nuclio/pkg/platform/kube"
+	"github.com/nuclio/nuclio/pkg/platform/kube/monitoring"
+	kubetest "github.com/nuclio/nuclio/pkg/platform/kube/test"
+
+	"github.com/stretchr/testify/suite"
+	appsv1 "k8s.io/api/apps/v1"
+	"k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+type FunctionMonitoringTestSuite struct {
+	kubetest.KubeTestSuite
+}
+
+func (suite *FunctionMonitoringTestSuite) SetupSuite() {
+	suite.KubeTestSuite.SetupSuite()
+
+	// decrease blocking interval, to make test run faster
+	monitoring.PostDeploymentMonitoringBlockingInterval = time.Second
+}
+
+func (suite *FunctionMonitoringTestSuite) TestNoRecoveryAfterBuildError() {
+	functionName := "function-build-fail"
+	createFunctionOptions := suite.CompileCreateFunctionOptions(functionName)
+	getFunctionOptions := &platform.GetFunctionsOptions{
+		Name:      createFunctionOptions.FunctionConfig.Meta.Name,
+		Namespace: createFunctionOptions.FunctionConfig.Meta.Namespace,
+	}
+
+	// fail during build
+	createFunctionOptions.FunctionConfig.Spec.Build.Commands = []string{
+		"exit 1",
+	}
+
+	// wait for at least one monitor + post deployment blocking intervals
+	waitOnMonitoringIntervals := 2*suite.Controller.GetFunctionMonitoringInterval() +
+		monitoring.PostDeploymentMonitoringBlockingInterval
+
+	suite.DeployFunctionExpectErrorAndRedeploy(createFunctionOptions,
+		func(deployResult *platform.CreateFunctionResult) bool {
+
+			// wait for monitoring
+			time.Sleep(waitOnMonitoringIntervals)
+
+			// ensure function is still in error state (due to build error)
+			suite.GetFunctionAndExpectState(getFunctionOptions, functionconfig.FunctionStateError)
+
+			// remove failing build command
+			createFunctionOptions.FunctionConfig.Spec.Build.Commands = []string{}
+
+			// function should be in error state
+			return true
+		}, func(deployResult *platform.CreateFunctionResult) bool {
+
+			// let interval occur at least once
+			time.Sleep(waitOnMonitoringIntervals)
+
+			// ensure function in ready state, build passes
+			suite.GetFunctionAndExpectState(getFunctionOptions, functionconfig.FunctionStateReady)
+
+			// make next deploy fail again
+			createFunctionOptions.FunctionConfig.Spec.Build.Commands = []string{
+				"exit 1",
+			}
+
+			suite.DeployFunctionExpectError(createFunctionOptions,
```

Comment:
the last deployment + recovery here seems to be redundant

Reply:
yup